File: python/day04.py
from aoc import *


def part_1(grid):
    letters = list(enumerate("XMAS"))
    deltas = (-1, 0, 1)
    return sum(1 for (x, y), c in grid.items()
               if c == 'X'
               for dx in deltas
               for dy in deltas
               if all(l == grid.get((x+i*dx, y+i*dy))
                      for i, l in letters))

def part_2(grid):
    diags = mapt(list, ("MAS", "SAM"))
    deltas = (-1, 0, 1)
    return sum(1 for x, y in grid.keys()
               if ([grid.get((x+d, y+d)) for d in deltas] in diags and
                   [grid.get((x-d, y+d)) for d in deltas] in diags))

def solve(filename):
    grid = list2grid(parse_lines(read_input(filename)))
    return part_1(grid), part_2(grid)


solve(4)

File: elixir/day04.exs
defmodule Day04 do
  defp parse_input(filename) do
    import AOC

    filename
    |> read_input
    |> parse_lines(:chars)
    |> grid_to_point_map
  end

  defp part_1(grid) do
    letters = Enum.with_index(String.graphemes("XMAS"))

    for {{x, y}, c} <- grid,
        c == "X",
        dx <- [-1, 0, 1],
        dy <- [-1, 0, 1],
        Enum.all?(
          letters,
          fn {l, i} ->
            l == Map.get(grid, {x + i * dx, y + i * dy})
          end
        ) do
      1
    end
    |> Enum.count()
  end

  defp check_diags(grid, x, y) do
    diags = [["M", "A", "S"], ["S", "A", "M"]]
    deltas = [-1, 0, 1]
    d1 = for d <- deltas, do: Map.get(grid, {x + d, y + d})
    d2 = for d <- deltas, do: Map.get(grid, {x - d, y + d})

    d1 in diags and d2 in diags
  end

  defp part_2(grid) do
    Map.keys(grid)
    |> Enum.count(fn {x, y} -> check_diags(grid, x, y) end)
  end

  def solve(filename) do
    filename
    |> parse_input
    |> AOC.juxt([&part_1/1, &part_2/1])
  end
end

Day04.solve(4)
